README.md
#Shape: A fluent HTTP Client for GWT (and Javascript soon!)

This library wraps GWT RequestBuilder into a fluent API.

```
Shape.post("http://httpbin.org/post")
  .queryString("name", "Mark")
  .field("last", "Polo")
  .asJson()
```

## Setup with Maven

Add dependency

```
<dependency>
    <groupId>com.dotweblabs.shape</groupId>
	<artifactId>shape</artifactId>
	<version>0-SNAPSHOT</version>
</dependency>
```

Add Bintray JCenter repository

```
<repositories>
    <repository>
	    <snapshots>
		    <enabled>false</enabled>
		</snapshots>
		<id>bintray-dotweblabs-maven</id>
		<name>bintray</name>
		<url>http://dl.bintray.com/dotweblabs/maven</url>
	</repository>
</repositories>
<pluginRepositories>
	<pluginRepository>
		<snapshots>
			<enabled>false</enabled>
		</snapshots>
		<id>bintray-dotweblabs-maven</id>
		<name>bintray-plugins</name>
		<url>http://dl.bintray.com/dotweblabs/maven</url>
	</pluginRepository>
</pluginRepositories>
```

CHANGELOG.md
## 0-SNAPSHOT - 04/18/17
- Fixed failing JUnit tests
- Added JUnit test delay
- Added Logging
- Updated JUnit version

## 0-SNAPSHOT - 01/30/17
- Add set timeout on requests

## 0-SNAPSHOT - 06/30/16
- Null check on RequestBuilder setHeader
- Header will not be added when value is null

## 0-SNAPSHOT - 06/29/16
- Fail on Request exception

## 0-SNAPSHOT - 06/27/16

- Remove Diamond operators

## 0-SNAPSHOT - 06/26/16

- Throw exception on response code greater than or equal to 400

## 0-SNAPSHOT - 06/24/16

### Fixed

- Wrong getModuleName()

### Added

- Initial commit


src/test/java/com/dotweblabs/shape/client/ShapeTest.java
/**
 *
 * Copyright (c) 2017 Dotweblabs Web Technologies and others. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.dotweblabs.shape.client;

import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONParser;
import com.google.gwt.json.client.JSONString;
import com.google.gwt.junit.client.GWTTestCase;
import com.google.gwt.user.client.rpc.AsyncCallback;
import org.junit.FixMethodOrder;
import org.junit.runners.MethodSorters;

import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * Unit tests of {@link ShapeTest}
 * @author Kerby Martino
 * @since 0-SNAPSHOT
 * @version 0-SNAPSHOT
 *
 */
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class ShapeTest extends GWTTestCase {

    static final Logger logger = Logger.getLogger(ShapeTest.class.getName());

    @Override
    public String getModuleName() {
        return "com.dotweblabs.shape.Shape";
    }

    public void test() {

    }

    public void testGet() {
        Shape.get("https://httpbin.org/get")
                .header("accept", "application/json")
                .header("Content-Type", "application/json")
                .asJson(new AsyncCallback<String>() {
                    @Override
                    public void onFailure(Throwable throwable) {
                        logger.log(Level.SEVERE, throwable.getMessage());
                        fail();
                    }
                    @Override
                    public void onSuccess(String s) {
                        logger.info(s);
                        JSONObject json = (JSONObject) JSONParser.parseStrict(s);
                        JSONString url = json.get("url").isString();
                        JSONObject headers = json.get("headers").isObject();
                        assertNotNull(json);
                        assertNotNull(url);
                        assertNotNull(headers);
                        log(s);
                        finishTest();
                    }
                });
        delayTestFinish(60000);
    }


    public void testPostFormField() {
        Shape.post("https://httpbin.org/post")
                .header("accept", "application/json")
                .header("Content-Type", "application/json")
                .queryString("name", "Mark")
                .field("middle", "O")
                .field("last", "Polo")
                .asJson(new AsyncCallback<String>() {
                    @Override
                    public void onFailure(Throwable throwable) {
                        logger.log(Level.SEVERE, throwable.getMessage());
                        fail();
                    }

                    @Override
                    public void onSuccess(String s) {
                        JSONObject json = (JSONObject) JSONParser.parseStrict(s);
                        JSONString url = json.get("url").isString();
                        JSONObject headers = json.get("headers").isObject();
                        JSONObject args = json.get("args").isObject();

                        assertNotNull(json);
                        assertNotNull(url);
                        assertNotNull(headers);
                        assertNotNull(args);

                        String name = args.get("name").isString().stringValue();
                        assertEquals("Mark", name);

                        log(s);
                        finishTest();

                    }
                });
        delayTestFinish(60000);
    }

    public void testPostJson() {
        JSONObject payload = new JSONObject();
        payload.put("hello", new JSONString("world"));

        Shape.post("https://httpbin.org/post")
                //.header("accept", "application/json")
                //.header("Content-Type", "application/json")
                .queryString("name", "Mark")
                .body(payload.isObject().toString())
                .asJson(new AsyncCallback<String>() {
                    @Override
                    public void onFailure(Throwable throwable) {
                        logger.log(Level.SEVERE, throwable.getMessage());
                        fail();
                    }

                    @Override
                    public void onSuccess(String s) {
                        JSONObject json = (JSONObject) JSONParser.parseStrict(s);
                        JSONString url = json.get("url").isString();
                        JSONObject headers = json.get("headers").isObject();
                        JSONObject jsonField = json.get("json").isObject();

                        assertNotNull(json);
                        assertNotNull(url);
                        assertNotNull(headers);
                        assertNotNull(jsonField);

                        String accept = headers.get("Accept").isString().stringValue();
                        String contentType = headers.get("Content-Type").isString().stringValue();
                        String hello = jsonField.get("hello").isString().stringValue();

                        assertEquals("application/json", accept);
                        assertEquals("application/json", contentType);
                        assertEquals("world", hello);

                        log(s);
                        finishTest();

                    }
                });
        delayTestFinish(60000);
    }

    public void testPut() {
        JSONObject payload = new JSONObject();
        payload.put("hello", new JSONString("world"));

        Shape.put("https://httpbin.org/put")
                .queryString("name", "Mark")
                .body(payload.isObject().toString())
                .basicAuth("john", "doe")
                .asJson(new AsyncCallback<String>() {
                    @Override
                    public void onFailure(Throwable throwable) {
                        logger.log(Level.SEVERE, throwable.getMessage());
                        fail();
                    }

                    @Override
                    public void onSuccess(String s) {
                        JSONObject json = (JSONObject) JSONParser.parseStrict(s);
                        JSONString url = json.get("url").isString();
                        JSONObject headers = json.get("headers").isObject();
                        JSONObject jsonField = json.get("json").isObject();

                        assertNotNull(json);
                        assertNotNull(url);
                        assertNotNull(headers);
                        assertNotNull(jsonField);

                        String accept = headers.get("Accept").isString().stringValue();
                        String authorization = headers.get("Authorization").isString().stringValue();
                        String contentType = headers.get("Content-Type").isString().stringValue();
                        String hello = jsonField.get("hello").isString().stringValue();

                        assertEquals("application/json", accept);
                        assertEquals("application/json", contentType);
                        assertEquals("world", hello);

                        String actual = "Basic " + Base64.btoa("john" + ":" + "doe");
                        assertEquals(actual, authorization);

                        log(s);
                        finishTest();

                    }
                });
        delayTestFinish(60000);
    }

    public void testDelete() {
        JSONObject payload = new JSONObject();
        payload.put("hello", new JSONString("world"));

        Shape.delete("https://httpbin.org/delete")
                .queryString("name", "Mark")
                .body(payload.isObject().toString())
                .basicAuth("john", "doe")
                .asJson(new AsyncCallback<String>() {
                    @Override
                    public void onFailure(Throwable throwable) {
                        logger.log(Level.SEVERE, throwable.getMessage());
                        fail();
                    }

                    @Override
                    public void onSuccess(String s) {
                        JSONObject json = (JSONObject) JSONParser.parseStrict(s);
                        JSONString url = json.get("url").isString();
                        JSONObject headers = json.get("headers").isObject();
                        //JSONObject jsonField = json.get("json").isObject();

                        assertNotNull(json);
                        assertNotNull(url);
                        assertNotNull(headers);
                        //assertNotNull(jsonField);

                        String accept = headers.get("Accept").isString().stringValue();
                        String authorization = headers.get("Authorization").isString().stringValue();
                        String contentType = headers.get("Content-Type").isString().stringValue();
                        //String hello = jsonField.get("hello").isString().stringValue();

                        assertEquals("application/json", accept);
                        assertEquals("application/json", contentType);
                        //assertEquals("world", hello);

                        String actual = "Basic " + Base64.btoa("john" + ":" + "doe");
                        assertEquals(actual, authorization);

                        log(s);
                        finishTest();

                    }
                });
        delayTestFinish(60000);
    }
    
    public static void log(String s){
        System.out.println(s);
    }

}


src/main/java/com/dotweblabs/shape/client/Shape.java
/**
 *
 * Copyright (c) 2017 Dotweblabs Web Technologies and others. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.dotweblabs.shape.client;

import com.google.gwt.http.client.RequestBuilder;

/**
 *
 * Fluent HTTP Client wrapper for GWT
 *
 * @author Kerby Martino
 * @since 0-SNAPSHOT
 * @version 0-SNAPSHOT
 */
public class Shape {
    public static GetRequest get(String url){
        return new GetRequest(url);
    };
    public static GetRequest head(String url){
        return new GetRequest(url);
    };
    public static HttpRequestWithBody post(String url){
        return new HttpRequestWithBody(url, RequestBuilder.POST);
    };
    public static HttpRequestWithBody put(String url){
        return new HttpRequestWithBody(url, RequestBuilder.PUT);
    };
    // TODO Not supported by RequestBuilder
    //public static HttpRequestWithBody patch(String url){
    //    return new HttpRequestWithBody(url, RequestBuilder.PATCH);
    //};
    //public static HttpRequestWithBody options(String url){
    //    return new HttpRequestWithBody(url, RequestBuilder.OPTIONS);
    //};
    public static HttpRequestWithBody delete(String url){
        return new HttpRequestWithBody(url, RequestBuilder.DELETE);
    };
}


src/main/java/com/dotweblabs/shape/client/Base64.java
/**
 *
 * Copyright (c) 2017 Dotweblabs Web Technologies and others. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.dotweblabs.shape.client;

/**
 *
 * @author Kerby Martino
 * @since 0-SNAPSHOT
 * @version 0-SNAPSHOT
 *
 */
public class Base64 {
    public static native String btoa(String a) /*-{
        return window.btoa(a);
    }-*/;
}


src/main/java/com/dotweblabs/shape/client/GetRequest.java
/**
 *
 * Copyright (c) 2017 Dotweblabs Web Technologies and others. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.dotweblabs.shape.client;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import com.google.gwt.http.client.*;
import com.google.gwt.user.client.rpc.AsyncCallback;

import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author Kerby Martino
 * @since 0-SNAPSHOT
 * @version 0-SNAPSHOT
 *
 */
public class GetRequest {

    static final Logger logger = Logger.getLogger(GetRequest.class.getName());

    private String url;
    private Multimap<String,String> headerMap;
    private Map<String, String> queryMap = null;
    private Map<String,Object> fields = null;

    private String authorization = null;
    private int TIMEOUT = 60000;

    public GetRequest(String url) {
        setUrl(url);
        headerMap = ArrayListMultimap.create();
    }

    public GetRequest header(String header, String value) {
        if(headerMap == null){
            headerMap = ArrayListMultimap.create();
        }
        if(value != null) {
            headerMap.put(header, value);
        }
        return this;
    }

    public GetRequest basicAuth(String username, String password) {
        authorization = "Basic " + Base64.btoa(username + ":" + password);
        return this;
    }

    public void asJson(final AsyncCallback<String> callback){
        if(queryMap != null && !queryMap.isEmpty()){
            url = url + "?";
            url = url +  queries(queryMap);
        }
        RequestBuilder b = new RequestBuilder(RequestBuilder.GET, url);
        b.setTimeoutMillis(TIMEOUT);
        if(headerMap != null){
            // Set default first
            headerMap.put("Content-Type", "application/json");
            headerMap.put("accept", "application/json");
            for (Map.Entry<String,String> entry : headerMap.entries()) {
                if(entry.getKey() != null && entry.getValue() != null
                        && !entry.getKey().isEmpty() && !entry.getValue().isEmpty()) {
                    b.setHeader(entry.getKey(), entry.getValue());
                }
            }
        }
        if(authorization != null){
            b.setHeader("Authorization", authorization);
        }
        try {
            b.sendRequest(null, new RequestCallback() {
                public void onResponseReceived(Request request, Response response) {
                    String resp = response.getText();
                    callback.onSuccess(resp);
                }
                public void onError(Request request, Throwable exception) {
                    callback.onFailure(exception);
                }
            });
        } catch (RequestException ex) {
            ex.printStackTrace();
            callback.onFailure(ex);
        }
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    private String queries(Map<String,String> parmsRequest){
        StringBuilder sb = new StringBuilder();
        for ( String k: parmsRequest.keySet() ) {
            String vx = URL.encodeComponent( parmsRequest.get(k));
            if ( sb.length() > 0 ) {
                sb.append("&");
            }
            sb.append(k).append("=").append(vx);
        }
        return sb.toString();
    }

    public void setTimeout(int timeout) {
        this.TIMEOUT = timeout;
    }

}


src/main/java/com/dotweblabs/shape/client/HttpRequestException.java
/**
 *
 * Copyright (c) 2017 Dotweblabs Web Technologies and others. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.dotweblabs.shape.client;

public class HttpRequestException extends Exception {
    private int code;
    public HttpRequestException(String message, int code){
        super(message);
        this.code = code;
    }
    public HttpRequestException() {
    }

    public int getCode() {
        return code;
    }
}


src/main/java/com/dotweblabs/shape/client/HttpRequestWithBody.java
/**
 *
 * Copyright (c) 2017 Dotweblabs Web Technologies and others. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.dotweblabs.shape.client;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import com.google.gwt.core.client.JavaScriptObject;
import com.google.gwt.http.client.*;
import com.google.gwt.user.client.rpc.AsyncCallback;

import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.logging.Logger;

/**
 *
 * @author Kerby Martino
 * @since 0-SNAPSHOT
 * @version 0-SNAPSHOT
 *
 */
public class HttpRequestWithBody {

    static final Logger logger = Logger.getLogger(HttpRequestWithBody.class.getName());

    private String url;
    private Multimap<String,String> headerMap;
    private Map<String, String> queryMap;
    private Map<String,Object> fields;
    private Object body = null;
    private RequestBuilder.Method method;
    private int TIMEOUT = 60000;

    private String authorization;

    public HttpRequestWithBody(String url, RequestBuilder.Method method) {
        setUrl(url);
        this.method = method;
        headerMap = ArrayListMultimap.create();
    }

    public HttpRequestWithBody header(String header, String value) {
        if(headerMap == null){
            headerMap = ArrayListMultimap.create();
        }
        if(value != null) {
            headerMap.put(header, value);
        }
        return this;
    }

    public HttpRequestWithBody body(Object body){
        this.body = body;
        return this;
    }

    public HttpRequestWithBody queryString(String name, String value){
        if(queryMap == null){
            queryMap = new LinkedHashMap<String,String>();
        }
        queryMap.put(name, value);
        return this;
    }

    public HttpRequestWithBody field(String name, String value){
        if(fields == null){
            fields = new LinkedHashMap<String,Object>();
        }
        fields.put(name, value);
        return this;
    }

    public HttpRequestWithBody basicAuth(String username, String password) {
        authorization = "Basic " + Base64.btoa(username + ":" + password);
        return this;
    }

    public void asJson(final AsyncCallback<String> callback){
        if(queryMap != null && !queryMap.isEmpty()){
            url = url + "?";
            url = url +  queries(queryMap);
        }
        RequestBuilder b = new RequestBuilder(method, url);
        b.setTimeoutMillis(TIMEOUT);
        if(headerMap != null){
            // Set default first
            headerMap.put("Content-Type", "application/json");
            headerMap.put("accept", "application/json");
            for (Map.Entry<String,String> entry : headerMap.entries()) {
                if(entry.getKey() != null && entry.getValue() != null
                        && !entry.getKey().isEmpty() && !entry.getValue().isEmpty()) {
                    b.setHeader(entry.getKey(), entry.getValue());
                }
            }
        }
        try {
            Object payload = body;
            if(fields != null && !fields.isEmpty()){
                StringBuilder sb = new StringBuilder();
                Iterator<Map.Entry<String,Object>> it = fields.entrySet().iterator();
                while(it.hasNext()){
                    Map.Entry<String,Object> entry = it.next();
                    if(entry.getValue() instanceof String){
                        if(!it.hasNext()){
                            sb.append(entry.getKey()).append("=").append(URL.encodeComponent((String.valueOf(entry.getValue()))));
                        } else {
                            sb.append(entry.getKey()).append("=").append(URL.encodeComponent((String.valueOf(entry.getValue())))).append("&");
                        }
                    }
                }
                payload = sb.toString();
                b.setHeader("Content-Type","application/x-www-form-urlencoded");
            }
            if(body != null){
                if(this.body instanceof JavaScriptObject){
                        // TODO for Sending File
                }
            }
            if(authorization != null){
                b.setHeader("Authorization", authorization);
            }
            if(payload != null) {
                b.sendRequest(String.valueOf(payload), new RequestCallback() {
                    public void onResponseReceived(Request request, Response response) {
                        String resp = response.getText();
                        if(response.getStatusCode() >= 400){
                            callback.onFailure(new HttpRequestException(resp, response.getStatusCode()));
                        } else {
                            callback.onSuccess(resp);
                        }
                    }
                    public void onError(Request request, Throwable exception) {
                        callback.onFailure(exception);
                    }
                });
            } else {
                b.sendRequest("", new RequestCallback() {
                    public void onResponseReceived(Request request, Response response) {
                        String resp = response.getText();
                        if(response.getStatusCode() >= 400){
                            callback.onFailure(new HttpRequestException(resp, response.getStatusCode()));
                        } else {
                            callback.onSuccess(resp);
                        }
                    }
                    public void onError(Request request, Throwable exception) {
                        callback.onFailure(exception);
                    }
                });
            }

        } catch (RequestException ex) {
            ex.printStackTrace();
            callback.onFailure(ex);
        }
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    private String queries(Map<String,String> parmsRequest){
        StringBuilder sb = new StringBuilder();
        for ( String k: parmsRequest.keySet() ) {
            String vx = URL.encodeComponent( parmsRequest.get(k));
            if ( sb.length() > 0 ) {
                sb.append("&");
            }
            sb.append(k).append("=").append(vx);
        }
        return sb.toString();
    }

    public void setTimeout(int timeout) {
        this.TIMEOUT = timeout;
    }

}
